src/compiler/checker.ts

```diff
@@ -2401,6 +2463,22 @@ namespace ts {
                         return getReturnTypeOfSignature(getSignatureFromDeclaration(getter));
                     }
                 }
+
+                // In a JS file, we might need to add types to parameters of a function expression that is
+                // an argument to 'define'
+                if (func.kind === SyntaxKind.FunctionExpression &&
+                    func.parent &&
+                    func.parent.kind === SyntaxKind.CallExpression) {
+                    
+                    if (isDefineCall(<CallExpression>func.parent) || isAmdRequireCall(<CallExpression>func.parent)) {
+                        assignDefineOrRequireCallParameterTypes(<FunctionExpression>func);

```

Comment:
This doesn't look right to me. You're basically *pushing* the type into the parameter instead of using *pulling* as we already do for contextual types. I suggest you move the check for a parameter of a callback of the AMD `define` function into `getContextuallyTypedParameterType` and then rely on existing machinery to perform the type assignment.